include/swift/AST/AutoDiff.h

```diff
@@ -79,6 +79,87 @@ class AutoDiffParameter {
 };
 
 class AnyFunctionType;
+class Type;
+
+/// Identifies a subset of a function's parameters.
+///
+/// Works with AST-level function decls and types. Requires further lowering to
+/// work with SIL-level functions and types. (In particular, tuples must be
+/// exploded).
+class AutoDiffParameterIndices {
+  /// Bits corresponding to parameters in the set are "on", and bits
+  /// corresponding to parameters not in the set are "off".
+  ///
+  /// When the function is curried, bits corresponding to all parameter groups
+  /// are flattened together in reverse order. For example,
```

Comment:
The Self parameter was the only thing I was using the reverse-flattening representation for. I can't think of any situation where it would be useful to do reverse-flattening for more general stuff.

Therefore, I have changed the API, comments, and implementation so that it only does 0 or 1 uncurryings depending on whether it's a method.